### PR TITLE
[Wire-Protocol 2/3 ] CoSi Changes

### PR DESCRIPTION
**Pull request by nikkolasg**
This second PR implements the `ProtocolIO` for the CoSi protocol; namely it creates **the packet**  `ProtocolPacket` for CoSi which contains everything:
- `Announcement`
- `Commitment`
- `Challenge`
- `Response`
- `OverlayMessage`

See #287

### CODE REVIEW

File: protocols/cosi/packets.go [hunk 1]
```diff
@@ -1,14 +1,104 @@
...
+
+// ProtocolPacketID is the network.PacketTypeID of the CoSi ProtocolPacket
+var ProtocolPacketID = network.RegisterPacketType(ProtocolPacket{})
+
+// ProtocolPacket is the main message for the CoSi protocol which includes
+// every information that the CoSi protocol might need.
+type ProtocolPacket struct {
+	Phase uint32
+
+	Info *sda.OverlayMessage
```
Comment (ineiti) [inline protocols/cosi/packets.go]: `Overlay` instead of `Info`?

File: protocols/cosi/packets.go [hunk 2]
```diff
@@ -1,14 +1,104 @@
...
+}
+
+// ProtocolIO implements the sda.ProtocolIO interface for the CoSi protocol.
+type ProtocolIO struct{}
 
-// The main messages used by CoSi
+// Wrap implements the sda.ProtocolIO interface by wrapping up any of the
+// four-step messages into a ProtooclPacket.
+func (p *ProtocolIO) Wrap(msg interface{}, info *sda.OverlayMessage) (interface{}, error) {
+	var packet = new(ProtocolPacket)
```
Comment (ineiti) [inline protocols/cosi/packets.go]: `overlay` instead of `info`?

File: services/cosi/cosi.go [hunk 3]
```diff
@@ -48,13 +48,13 @@ type SignatureResponse struct {
 // SignatureRequest treats external request to this service.
 func (cs *CoSi) SignatureRequest(si *network.ServerIdentity, req *SignatureRequest) (network.Body, error) {
 	tree := req.Roster.GenerateBinaryTree()
-	tni := cs.NewTreeNodeInstance(tree, tree.Root, protocol.Name)
-	pi, err := protocol.NewCoSi(tni)
+	tni := cs.NewTreeNodeInstance(tree, tree.Root, cosi.Name)
+	pi, err := cosi.NewCoSi(tni)
```
Comment (ineiti) [inline services/cosi/cosi.go]: Perhaps now that it's called `cosi` we should rename `NewCosi`, so it's more like `cosi.NewProtocol`?

Comment (nikkolasg) [inline services/cosi/cosi.go]: Yep agreed.

File: protocols/cosi/packets.go (hunk 1 above)
Comment (nikkolasg) [inline protocols/cosi/packets.go]: Mhhh, I'd like the name to represent that it is only a packet, a name, a set of information received, and not a structure which can *do* things such as `Overlay`, so `OverlayMessage` maybe ?

Comment (ineiti) [inline protocols/cosi/packets.go]: +1 for `OverlayMessage`

**Pull request merged**